### ISSUE TITLE
Shadow relative error

### ISSUE THREAD

**Issue by marchdf** [manufactured-solutions/MASA#46]
@nicholasmalaya posting here per our email conversation: 

I am working with Jeremy Melvin and Bob Moser on some MMS problems and I have been using MASA to add LES models to the CNS in order to do MMS for LES in our project (here's the branch if you care to look at it: https://github.com/marchdf/MASA/tree/add-cns-les). I am hoping to get things in order so that I can eventually merge it into MASA proper and so others on our team can use it as well. In view of doing so, I added a test to my MASA branch (https://github.com/marchdf/MASA/blob/add-cns-les/tests/ad_cns_les.cpp). The test passes but it also spews out a bunch of `Shadow relative error = 3.30261e-05` that pollutes the `make check` output.  I noticed some of the other tests do this but to a way lesser degree (maybe 4 or 5 of those, whereas for me, it fills up the screen). I can get rid of it by using `typedef double RawScalar` instead of `typedef ShadowNumber<double, long double> RawScalar;` but I doubt that's the right way to do so. What is this from? Should I worry about it? Should I leave it as is or do something to get rid of it?

**Comment by nicholasmalaya** [comment]
Can you post the error log? And are the tests passing in make check? Finally, can you give me a snapshot of the machine you are running on, in terms of hardware?

@roystgnr please correct me if you disagree, this is more your wheelhouse.

**Comment by roystgnr** [comment]
> What is this from?

The "shadow number" class runs every computation in parallel, with two different data types, and screams at you if the relative error between the results is more than an order of magnitude greater than the machine precision of the first datatype.  The goal is to find out whether your computation is particularly sensitive to roundoff errors.

> Should I worry about it?

Maybe?

3e-5 is dumbfoundingly huge, as roundoff errors go.  That's not a bit more than an order of magnitude times double epsilon, that's like ten orders of magnitude times double epsilon.  If I saw that kind of relative error while testing against a manufactured solution, I would assume something was wrong with my code, not the solution.

However, the question of *where* you're getting the error is important too.  If you have a PDE solution that nears or crosses zero, then around those points the relative error can easily go to infinity.  When testing your PDE you'll use an absolute error tolerance to avoid false positives in those regions, but ShadowNumber doesn't currently have any facilities for that.

> Should I leave it as is or do something to get rid of it?

Let me see if I can replicate it, and if so whether I can thereby understand it better.

**Comment by marchdf** [comment]
- [Attached is the output from `make check` where you can see all the `Shadow relative` lines.](https://github.com/manufactured-solutions/MASA/files/1142301/out.txt).
- All the tests pass in `make check`
- I am running on NREL supercomputer called Peregrine with Intel Haswell nodes, Linux CentOS 6, gcc 5.2.

**Comment by roystgnr** [comment]
Wow.  Forget 3e-5, you've got relative errors there of 3e-1!

**Comment by marchdf** [comment]
@roystgnr Thank you for those explanations! Seems like something seriously wrong is going on...

**Comment by roystgnr** [comment]
No, looks like false positives.  You're taking the symmetric part of an extremely anti-symmetric tensor.

```
(gdb) p GradU[0][2].value().value()
$47 = (double &) @0x7ffffffde310: -251769.18872456613
(gdb) p GradU[2][0].value().value()
$48 = (double &) @0x7ffffffde550: 251769.18872456608
(gdb) p S[0][2].value().value()
$49 = (MetaPhysicL::ShadowNumber<double, long double> &) @0x7ffffffdd180: {<MetaPhysicL::safe_bool<MetaPhysicL::ShadowNumber<double, long double> >> = {<MetaPhysicL::safe_bool_base> = {<No data fields>}, <No data fields>}, _val = -2.9103830456733704e-11, 
  _shadow = -1.1155520951433572918176651000976562e-11}
```

If you actually cared about the relative error in those scalars then yeah, you'd be in big trouble with a relative error of 160%!  But you care about the error relative to the tensor as a whole, which is about 1e-16, which is fine.

You either need to propagate ShadowNumber through the whole calculation *without* using raw_value to strip it down to a double prematurely (the best solution) or you need to just use double and comment out the use of ShadowNumber (the quickest solution).

**Comment by roystgnr** [comment]
Leaving this issue open in case you do want to wrestle with ShadowNumber and you need more help; otherwise we can probably close it.

**Comment by marchdf** [comment]
> You either need to propagate ShadowNumber through the whole calculation without using raw_value 

I like this idea. A couple questions: 
1. Would I just have to replace `raw_value(*)` calls in the test file `tests/ad_cns_les.cpp`? Or would I have to do that in my src file as well? 
2. What would I replace them by?

**Comment by roystgnr** [comment]
The warnings I was tracking down all came from the test file, but you definitely want to get rid of as many intermediate raw_value calls as you can in the source file too.

Typically you'd replace them by nothing.  If "A = B" compiles, it's better than "A = raw_value(B)".  Even if not, you can often get away with e.g. changing the type of A to hold the non-raw information, or using "A = raw_value(B + C + D)" rather than "A = raw_value(B) + raw_value(C) + raw_value(D)" to postpone the shadow test until the last possible moment.

**Comment by marchdf** [comment]
Ok let me give that a shot later today. Will keep this open until I have this figured out if that is ok.

**Comment by roystgnr** [comment]
Sure

**Comment by marchdf** [comment]
So I think it was a combination of the abuse of `raw_value` and the solution itself. I removed the use of `raw_value` and that still left a couple `shadow relative errors` around 1e-5. Then I slightly modified the frequencies in my cosine solutions. Now the [output of `make check`](https://github.com/manufactured-solutions/MASA/files/1143349/out.txt) has a bunch of `shadow relative errors` which around 1e-14 or 1e-15. All the tests pass, however, there is still a lot of pollution from the `shadow relative errors` in that `make check` output. Should I worry about it or just leave it as is? Eventually I would like to merge this with MASA proper so I want to be sure I am doing the right thing...

**Comment by roystgnr** [comment]
1e-14 I wouldn't worry about, at all.  At this point the next fix isn't to change anything in your code, it's to make ShadowNumber more flexible about relative and absolute tolerances.

**Comment by marchdf** [comment]
Great, thank you for your help. I will close this now.
